### PR TITLE
fix(storage): set content-range headers correctly for streaming sources of unknown length

### LINKED ISSUE #26: [storage upload]: supported upload sources
- [ ] fixed buffers in memory
- [ ] files from disk
- [ ] dynamically created via computation
- [ ] data received asynchronously from external source (e.g. download)

### PR DESCRIPTION
**Pull request by chingor13**
Also adds tests for all our supported upload source types

Fixes #26

### CODE REVIEW

File: packages/storage/Tests/ResumableUploadTests.swift [hunk 1]
```diff
@@ -761,38 +761,76 @@ import Testing
...
+    #expect(object.bucket == bucket)
+    #expect(object.size == Int64(data.count))
+  }
+
+  /// Tests resuming an interrupted upload for a fixed buffer in memory (`BytesSource`).
+  @Test func resumeUploadFixedBufferMemorySuccess() async throws {
+    let registry = MockRegistry.create()
+    let bucket = "test-bucket"
+    let objectName = "fixed-buffer-resumed"
+    let data = Data(repeating: 0xCD, count: 10 * 1024 * 1024)  // 10MB
```
Comment (coryan) [inline packages/storage/Tests/ResumableUploadTests.swift]:
```suggestion
    let data = Data(repeating: 0xCD, count: 10 * 1024 * 1024)  // 10MiB
```

File: packages/storage/Tests/ResumableUploadTests.swift [hunk 2]
```diff
@@ -761,38 +761,76 @@ import Testing
...
+  // --- Source Type 1: Fixed Buffers in Memory ---
+
+  /// Tests resumable upload using a fixed buffer in memory (`BytesSource`).
+  @Test func resumableUploadFixedBufferMemorySuccess() async throws {
     let registry = MockRegistry.create()
     let bucket = "test-bucket"
-    let objectName = "test-object"
-    let data = Data(repeating: 1, count: 1 * 1024 * 1024)  // 1MB payload
+    let objectName = "fixed-buffer-object"
+    let data = Data(repeating: 0xAB, count: 10 * 1024 * 1024)  // 10MB
```
Comment (coryan) [inline packages/storage/Tests/ResumableUploadTests.swift]: nit:

```suggestion
    let data = Data(repeating: 0xAB, count: 10 * 1024 * 1024)  // 10MiB
```

File: packages/storage/Tests/ResumableUploadTests.swift [hunk 3]
```diff
@@ -761,38 +761,76 @@ import Testing
...
+
+    // GCS reports 5MB uploaded (bytes 0-5242879 received)
+    let offset: Int64 = 5 * 1024 * 1024
+    let lastByte = offset - 1
 
     registry.register(
       response: .success(
         statusCode: 308, data: Data(),
-        headers: ["Range": "bytes=0-4999"]),
+        headers: ["Range": "bytes=0-\(lastByte)"]),
```
Comment (coryan) [inline packages/storage/Tests/ResumableUploadTests.swift:829]: Please add some test cases where the query returns "no bytes are committed"...

File: packages/storage/Sources/GoogleCloudStorage/StorageClient+Upload.swift [hunk 4]
```diff
@@ -470,9 +489,13 @@ extension HTTPClient {
...
     request.applyCustomerSuppliedEncryptionHeaders(options.customerEncryptionKey)
 
-    let end = offset + Int64(data.count) - 1
     let totalStr = totalSize.map { String($0) } ?? "*"
-    request.setValue("bytes \(offset)-\(end)/\(totalStr)", forHTTPHeaderField: "Content-Range")
+    if data.isEmpty {
+      request.setValue("bytes */\(totalStr)", forHTTPHeaderField: "Content-Range")
+    } else {
+      let end = offset + Int64(data.count) - 1
+      request.setValue("bytes \(offset)-\(end)/\(totalStr)", forHTTPHeaderField: "Content-Range")
```
Comment (coryan) [inline packages/storage/Sources/GoogleCloudStorage/StorageClient+Upload.swift:496]: Add a test case where the data size is unknown, but turns out to be 0 bytes (an empty stream). How would you format this header in that case?

Also, consider a test case where the last chunk is empty.

**Review by coryan — CHANGES_REQUESTED** [review]
I think you have a bug for the empty stream case. Also streams that are exactly a multiple of the chunk size can be troublesome.

File: packages/storage/Tests/ResumableUploadTests.swift (hunk 3 above)
Comment (chingor13) [reply packages/storage/Tests/ResumableUploadTests.swift:829]: We have a `resumeZeroBytesOnServer()` test. In this case, the query response does not return a `Range` header

File: packages/storage/Sources/GoogleCloudStorage/StorageClient+Upload.swift (hunk 4 above)
Comment (chingor13) [reply packages/storage/Sources/GoogleCloudStorage/StorageClient+Upload.swift:496]: Added tests for dynamic content where both last chunk is empty and a completely empty stream. Also verified with an integration test that succeeds for both these scenarios.

File: packages/storage/Tests/ResumableUploadTests.swift (hunk 3 above)
Comment (coryan) [reply packages/storage/Tests/ResumableUploadTests.swift:829]: Ack, thanks.

**Review by coryan — APPROVED** [review]
